roaring/src/bitmap/store/bitmap_store.rs

```diff
@@ -403,6 +409,146 @@ impl Display for Error {
+            let block_iter =
+                Self::partial_block_iter(start_key, bits[start_key], start_bit..=end_bit);
+            return BlockRangeIter::single_block(block_iter);
+        }
+
+        let first = Self::partial_block_iter(start_key, bits[start_key], start_bit..=63);
+        let start_key_p1 = start_key + 1;
+        let between = BlockSeqIter::new(start_key_p1, &bits[start_key_p1..end_key]);
+        let last = Self::partial_block_iter(end_key, bits[end_key], 0..=end_bit);
+
+        BlockRangeIter { first, between, last }
+    }
+    #[inline]
+    fn start_mask(s: usize) -> u64 {
+        match s {
+            0 => u64::MAX,
+            _ if s >= 64 => u64::MIN,
+            _ => !((1u64 << s) - 1),
+        }
+    }
+    #[inline]
+    fn end_mask(e: usize) -> u64 {
+        match e {
+            _ if e >= 63 => u64::MAX,
+            _ => (1u64 << e + 1) - 1,
+        }
+    }
+    fn partial_block_iter(key: usize, block: u64, range: RangeInclusive<usize>) -> BlockPartIter {
+        let (start, end) = (*range.start(), *range.end());
+        let start_mask = Self::start_mask(start);
+        let end_mask = Self::end_mask(end);
+        BlockPartIter::new(key, block & start_mask & end_mask)
+    }
+    fn empty() -> BlockRangeIter<'a> {
+        BlockRangeIter {
+            first: BlockPartIter::empty(),
+            between: BlockSeqIter::empty(),
+            last: BlockPartIter::empty(),
+        }
+    }
+    fn single_block(block: BlockPartIter) -> BlockRangeIter<'a> {
+        BlockRangeIter {
+            first: block,
+            between: BlockSeqIter::empty(),
+            last: BlockPartIter::empty(),
+        }
+    }
+}
+
+struct BlockPartIter {
+    key: usize,
+    value: u64,
+}
+
+impl BlockPartIter {
+    fn new(key: usize, block: u64) -> BlockPartIter {
+        BlockPartIter { key, value: block }
+    }
+    fn empty() -> BlockPartIter {
+        Self::new(0, 0)
+    }
+}
+
+impl Iterator for BlockPartIter {
+    type Item = u16;
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.value == 0 {
+            return None;
+        }
+        let index = self.value.trailing_zeros() as usize;
+        self.value &= self.value - 1;
+        return Some((64 * self.key + index) as u16);
+    }
+}
+
+struct BlockSeqIter<'a> {
+    start_key: usize,
+    block_range: &'a [u64],
+}
+impl<'a> BlockSeqIter<'a> {
+    fn new(start_key: usize, block_range: &'a [u64]) -> BlockSeqIter<'a> {
+        BlockSeqIter { start_key, block_range }
+    }
+    fn empty() -> BlockSeqIter<'a> {
+        BlockSeqIter { start_key: 0, block_range: &[] }
+    }
+}
+impl<'a> Iterator for BlockSeqIter<'a> {
+    type Item = u16;
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.block_range.len() == 0 {
+            return None;
+        }
+        let mut current = BlockPartIter::new(self.start_key, self.block_range[0]);
+        if let c @ Some(_) = current.next() {
+            return c;
+        }
```

Comment:
Won't this always return the same value from `next` if there's a bit set in the "current" block, since we always reload `self.block_range[0]`? `self` is not modified in any way if we `return` here, right?